![](https://upload.wikimedia.org/wikipedia/en/b/bb/Titanic_breaks_in_half.jpg)

# Project 1: [Titanic](https://www.kaggle.com/c/titanic/data)
---

The sinking of the RMS Titanic is one of the most infamous shipwrecks in history.  On April 15, 1912, during her maiden voyage, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. This sensational tragedy shocked the international community and led to better safety regulations for ships.

One of the reasons that the shipwreck led to such loss of life was that there were not enough lifeboats for the passengers and crew. Although there was some element of luck involved in surviving the sinking, some groups of people were more likely to survive than others, such as women, children, and the upper-class.

In this project, you will exercise your skills with loading data, python data structures, and Pandas to identify charactaristics of Titanic survivors!

---
#### Your goals should be to:
* Practice python programming including loops, conditionals, types, functions, and data structures
* Start thinking critically about manipulating, organizing, and interpreting data
* Troubleshoot errors

---
#### Getting Started:
* **fork** the repository on git.generalassemb.ly
* **clone** your forked repo

---
#### Submission:
* You should be working on a **fork** of the GA project one repository. 
* Use **git** to manage versions of your project. Make sure to `add`, `commit`, and `push` your changes to **your fork** of the github 
* Submit a link to your project repository in the submission form by **Friday, 9/29 11:59 PM**. You will then receive the solutions.
* Create a copy of your original notebook (file > make a copy in jupyter notebook)
* In the copy, use the solutions to correct your work. Make sure to take note of your successes and struggles. Did you learn anything new from correcting your work?
* Submit the corrected version by **Sunday, 10/1 11:59 PM** to receive instructor feedback on your work. ***Projects submitted after this deadline will not receive instructor feedback.***

### Considerations:

* You will be generating long data strutures- avoid displaying the whole thing. Display just the first or last few entries and look at the length or shape to check whether your code gives you back what you want and expect.
* Make functions whenever possiblle!
* Be explicit with your naming. You may forget what `this_list` is, but you will have an idea of what `passenger_fare_list` is. Variable naming will help you in the long run!
* Don't forget about tab autocomplete!
* Use markdown cells to document your planning, thoughts, and results. 
* Delete cells you will not include in your final submission
* Try to solve your own problems using this framework:
  1. Check your spelling
  2. Google your errors. Is it on stackoverflow?
  3. Ask your classmates
  4. Ask a TA or instructor
* Do not include errors or stack traces (fix them!)

# 1. Using the `with open()` method in the `csv` library, load the titanic dataset into a list of lists.

* The `type()` of your dataset should be `list`
* The `type()` of each element in your dataset should also be `list`
* The `len()` of your dataset should be 892 (892 rows, including the header)
* The `len()` of each row element in your dataset should be have a `len()` of 12
* Print out the first 3 rows including the header to check your data.

In [56]:
import csv
from IPython.display import display
import numpy as np

In [57]:
with open('titanic.csv') as titanic:
    csv = csv.reader(titanic)
    titanic_list = list(csv)

print(titanic_list[0:3])

[['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], ['1', '0', '3', 'Braund, Mr. Owen Harris', 'male', '22', '1', '0', 'A/5 21171', '7.25', '', 'S'], ['2', '1', '1', 'Cumings, Mrs. John Bradley (Florence Briggs Thayer)', 'female', '38', '1', '0', 'PC 17599', '71.2833', 'C85', 'C']]


# 2. Separate the first header row from the rest of your dataset. 

* The header should be a list of the column names
* The data should be the rest of your data
* Display the header and the first row of the dataset zipped together using `zip`
* Your result should look like...


```
[('PassengerId', '1'),
 ('Survived', '0'),
 ('Pclass', '3'),
 ...
 ('Embarked', 'S')]
 ```

In [58]:
header = titanic_list[0]
data = titanic_list[1:]
first_zip = zip(header, data[0])

print(list(first_zip))

[('PassengerId', '1'), ('Survived', '0'), ('Pclass', '3'), ('Name', 'Braund, Mr. Owen Harris'), ('Sex', 'male'), ('Age', '22'), ('SibSp', '1'), ('Parch', '0'), ('Ticket', 'A/5 21171'), ('Fare', '7.25'), ('Cabin', ''), ('Embarked', 'S')]


# 3. Using a `for` loop, load your data into a `dict` called `data_dict`.

* The keys of your `data_dict` should be `PassengerId`
* The values of your `data_dict` should be dictionaries...
  * Each of these dictionaries should reperesent a column value within a row
  * The keys should be the names of the columns
  * The values should be the values of that column
  
The beginning of your `data_dict` should look like: 

    {'1': {'Age': '22',
      'Cabin': '',
      'Embarked': 'S',
      'Fare': '7.25',
      'Name': 'Braund, Mr. Owen Harris',
      'Parch': '0',
      'Pclass': '3',
      'Sex': 'male',
      'SibSp': '1',
      'Survived': '0',
      'Ticket': 'A/5 21171'},
     '10': {'Age': '14',
      'Cabin': '',
      'Embarked': 'C',
      'Fare': '30.0708',
      'Name': 'Nasser, Mrs. Nicholas (Adele Achem)',
      'Parch': '0',
      'Pclass': '2',
      'Sex': 'female',
      'SibSp': '1',
      'Survived': '1',
      'Ticket': '237736'},
      ...
      }

In [59]:
data_dict = {}
for j in range(len(data)):
    for a, b in zip(header ,data[j]):
        if a == 'PassengerId':
            c = b
            data_dict[c] = {}
        else:
            data_dict[c].update({a:b})
    
print(data_dict['1'])

{'Survived': '0', 'Pclass': '3', 'Name': 'Braund, Mr. Owen Harris', 'Sex': 'male', 'Age': '22', 'SibSp': '1', 'Parch': '0', 'Ticket': 'A/5 21171', 'Fare': '7.25', 'Cabin': '', 'Embarked': 'S'}


# 4. Repeat step 3 using a dictionary comprehension.

* Using `==`, check if your `data_dict` from your `for` loop is the same as the one from your dictionary comprehension.

In [3]:
data_dict_comp = {}

# 5. Transform your `data_dict` to be oriented by column and call it `data_dict_columns`

* Currently, our `data_dict` is oriented by row, indexed by `"PassengerId"`. 
* Transform your data so that the title of each row is a key, the values are of type `list` and represent column vectors.

If you display `data_dict_columns`, the beginning should look like...

    {'Age': ['25',
      '36',
      '24',
      '40',
      '45',
      '2',
      '24',
      '28',
      '33',
      '26',
      '39',
      ...

In [ ]:
from collections import defaultdict

data_dict_columns = defaultdict(list)
for j in data_dict:
    for a in data_dict[j]:
        data_dict_columns[a].append(data_dict[j][a])
print(data_dict_columns['Survived'])

# 6. Data Types

What is the current `type` of each column? What do you think the data type of each column *should* be? The data types in Python are...

* `int`
* `float`
* `str`
* `bool`
* `tuple`
* `list`
* `dict`
* `set`

In a markdown cell, describe what each column represents and what the `type` of each value should be. **Extra:** If you want to be fancy, use a [markdown table](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables) to display your results.

PassengerId - Represents an assigned PassengerId numbered 1 through 892. This should be an integer.

Survived - If the corresponding passenger survived, Survived = 1. If they did not survive, Survived = 0. This should also be an integer.

PClass - The socioeconomic class that the passenger belonged to, upper class = 1, middle class = 2, lower class = 3. This should also be represented by an integer.

Name - The name and title of the passenger. This should be represented by a string.

Sex - The sex of the passenger. This should be represented by a string.

Age - The age of the passenger. This should be represented by a integer.

SibSp - This column represents the number of siblings and/or spouses aboard the Titanic with the corresponding passenger. This should be represented by an integer.

Parch - Represents the number of parents and/or children aboard the Titanic with the corresponding passenger. This should also be represented by an integer.

Ticket - The ticket number of the passenger's ticket. Although called a ticket number, this should be represented as a string since the ticket numbers include both letters and numbers.

Fare - Amount of money the passenger spent on their ticket aboard the Titanic. Should be represented as a float object.

Cabin - The cabin number where the passenger was staying aboard the Titanic. The cabin number includes both letters and numbers and, therefore, should be represented as a string.

Embarked - The port that the passenger embarked from. This is represented as a single upper-case letter, therefore, should be represented as a string.

# 7. Transform each column to the appropriate type if needed.

Build a function called `transform_column` that takes arguments for a `data_dict`, `column_name`, and `datatype`, and use it to transofm the columns that need transformation.

**NOTE:** There are values in this dataset that cannot be directly cast to a numerical value. Use `if/then` or `try/except` statements to handle errors. 

**To help identify potential sources of errors, explore the `set` of values in each column.**

In [61]:
def transform_column(data_dict, column_name, datatype):
    new_data_dict = {}
    for j in data_dict:
        if j == column_name:
            new_data_dict[j] = []
            for a in data_dict[j]:
                try:
                    if a == '':
                        pass
                    else:
                        new_data_dict[j].append(datatype(a))
                except:
                    new_data_dict[j].append(a)
                
    for l in new_data_dict:
        if l == column_name:
            data_dict[l] = new_data_dict[l]
    return data_dict

transform_column(data_dict_columns, 'Survived', int)
transform_column(data_dict_columns, 'Age', float)
transform_column(data_dict_columns, 'SibSp', int)
transform_column(data_dict_columns, 'Parch', int)
transform_column(data_dict_columns, 'Pclass', int)
transform_column(data_dict_columns, 'Fare', float)

defaultdict(list,
            {'Age': [22.0,
              38.0,
              26.0,
              35.0,
              35.0,
              54.0,
              2.0,
              27.0,
              14.0,
              4.0,
              58.0,
              20.0,
              39.0,
              14.0,
              55.0,
              2.0,
              31.0,
              35.0,
              34.0,
              15.0,
              28.0,
              8.0,
              38.0,
              19.0,
              40.0,
              66.0,
              28.0,
              42.0,
              21.0,
              18.0,
              14.0,
              40.0,
              27.0,
              3.0,
              19.0,
              18.0,
              7.0,
              21.0,
              49.0,
              29.0,
              65.0,
              21.0,
              28.5,
              5.0,
              11.0,
              22.0,
              38.0,
              45.0,
              4.0,
   

# 8. Build functions to calculate the mean, sample standard deviation, and median of a list of ints or floats. Use `scipy.stats.mode` or build your own mode function!


If you filled any missing values with `np.NaN`, you may need to handle that in your functions (look up `np.isnan()`).

If building a `mode` function is too difficult, you import mode from `scipy.stats` using `from scipy.stats import mode`.

**Optional:**  Build a function for calculating the Mode that returns the mode value *and* the count of that value. Mode is tricky, so start by building a function that counts the occurances of each value. You may also need to sort using a `key` with a `lambda function` inside. You may also find a `defaultdict` useful.

Mean

In [62]:
def this_mean(data_list):
    total = 0
    mean = 0
    new_data_list = []
    for length in range(len(data_list)):
        if data_list[length] == '':
            pass
        else:
            new_data_list.append(data_list[length])
    for length in range(len(new_data_list)):
        total += new_data_list[length]
    return total/len(new_data_list)

Standard Deviation

In [63]:
def this_std(data_list):
    if data_list == data_dict_columns['Fare']:
        transform_column(data_dict_columns, data_list, float)
    else:
        transform_column(data_dict_columns, data_list, int)
    deviation_value = 0
    standard_deviation = 0
    length_of_list = len(data_list)
    for length in range(len(data_list)):
        if data_list[length] =='':
            length_of_list -= 1
        else:
            deviation_value += ((this_mean(data_list) - data_list[length]) ** 2)
    standard_deviation = np.sqrt(deviation_value/(length_of_list))
    return standard_deviation

Median

In [64]:
def this_median(data_list):
    new_data_list = sorted(data_list)
    return new_data_list[round(len(new_data_list)/2)]

Mode

In [65]:
def this_mode(data_list):
    count = []
    for j in range(len(data_list)):
        incrementer = 0
        for i in range(len(data_list)):
            if data_list[j] == data_list[i]:
                incrementer += 1
                count.append(incrementer)
            else:
                pass
    actual_count = 0
    keep = 0
    for u in range(len(count)):
        if count[u] > actual_count:
            actual_count = count[u]
            keep = u
    return data_list[keep], actual_count
this_mode(data_dict_columns['Survived'])
# SibSp, Age throws an error

(0, 549)

# 9. Summary Statistics of Numerical Columns

For numerical columns, what is the mean, standard deviation, mean, and mode for that data? Which measure of central tendency is the most descriptive of each column? Why? Explain your answer in a markdown cell.

In [66]:
print("The average age aboard the Titanic was {} years old".format(this_mean(data_dict_columns['Age'])))
print("The standard deviation of age aboard the Titanic was {} years".format(this_std(data_dict_columns['Age'])))
print("The median age aboard the Titanic was {} years old".format(this_median(data_dict_columns['Age'])))
# print("The mode age aboard the Titanic was {} years old".format(this_mode(data_dict_columns['Age'])))
print('\n')
print("The survival rate aboard the Titanic was {}".format(this_mean(data_dict_columns['Survived'])))
print("The standard deviation of the survival rate aboard the Titanic was {}".format(this_std(data_dict_columns['Survived'])))
print("The median survival (1 or 0) aboard the Titanic was {}".format(this_median(data_dict_columns['Survived'])))
print("The mode survival (1 or 0) aboard the Titanic was {} which appeared {} times".format(this_mode(data_dict_columns['Survived'])[0], this_mode(data_dict_columns['Survived'])[1]))
print('\n')
print("The average socioeconomic class aboard the Titanic was {}".format(this_mean(data_dict_columns['Pclass'])))
print("The standard deviation of socioeconomic class aboard the Titanic was {}".format(this_std(data_dict_columns['Pclass'])))
print("The median socioeconomic class aboard the Titanic was {}".format(this_median(data_dict_columns['Pclass'])))
print("The mode socioeconomic class aboard the Titanic was {} and it appeared {} times".format(this_mode(data_dict_columns['Pclass'])[0], this_mode(data_dict_columns['Pclass'])[1]))
print('\n')
print("The average number of siblings and/or spouses that each passenger had aboard the Titanic was {}".format(this_mean(data_dict_columns['SibSp'])))
print("The standard deviation of number of siblings and/or spouses aboard the Titanic was {}".format(this_std(data_dict_columns['SibSp'])))
print("The median number of siblings and/or spouses aboard the Titanic for each passenger was {}".format(this_median(data_dict_columns['SibSp'])))
# print("The mode number of siblings and/or spouses aboard the Titanic for each passenger was {}".format(this_mode(data_dict_columns['SibSp'])))
print('\n')
print("The average number of parents and/or children that each passenger had aboard the Titanic was {}".format(this_mean(data_dict_columns['Parch'])))
print("The standard deviation of parents and/or children that each passenger had aboard the Titanic was {}".format(this_std(data_dict_columns['Parch'])))
print("The median number of parents and/or children aboard the Titanic for each passenger was {}".format(this_median(data_dict_columns['Parch'])))
print("The mode number of parents and/or children aboard the Titanic for each passenger was {}".format(this_mode(data_dict_columns['Parch'])[0]))
print('\n')
print("The average ticket fare was ${}".format(this_mean(data_dict_columns['Fare'])))
print("The standard deviation of ticket fare was ${}".format(this_std(data_dict_columns['Fare'])))
print("The median ticket fare was ${}".format(this_median(data_dict_columns['Fare'])))
print("The mode ticket fare was ${} and it appeared {} times".format(this_mode(data_dict_columns['Fare'])[0], this_mode(data_dict_columns['Fare'])[1]))







The average age aboard the Titanic was 29.69911764705882 years old
The standard deviation of age aboard the Titanic was 14.516321150817317 years
The median age aboard the Titanic was 28.0 years old


The survival rate aboard the Titanic was 0.3838383838383838
The standard deviation of the survival rate aboard the Titanic was 0.48631931786710086
The median survival (1 or 0) aboard the Titanic was 0
The mode survival (1 or 0) aboard the Titanic was 0 which appeared 549 times


The average socioeconomic class aboard the Titanic was 2.308641975308642
The standard deviation of socioeconomic class aboard the Titanic was 0.8356019334795189
The median socioeconomic class aboard the Titanic was 3
The mode socioeconomic class aboard the Titanic was 3 and it appeared 491 times


The average number of siblings and/or spouses that each passenger had aboard the Titanic was 0.5230078563411896
The standard deviation of number of siblings and/or spouses aboard the Titanic was 1.1021244350892836
The med

Your markdown response here

# 10. Splitting the Data to Predicting Survival

For all the passengers in the dataset, the mean survival rate is around .38 (38% of the passengers survived). From our data, we may be able to profile who survived and who didn't!

Split the data by pclass. Does the class a passenger was in affect survivability? You can do this by:
* Creating a list of `True` and `False` values conditional on a column's value
* Taking the mean of the `Survived` column where those values are `True`

In [67]:
pclass1_survived = 0
pclass1_incrementer = 0
pclass2_survived = 0
pclass2_incrementer = 0
pclass3_survived = 0
pclass3_incrementer = 0
for r in range(len(data_dict_columns['Pclass'])):
    if data_dict_columns['Pclass'][r] == 1:
        pclass1_incrementer += 1
        if data_dict_columns['Survived'][r] == 1:
            pclass1_survived += 1
    if data_dict_columns['Pclass'][r] == 2:
        pclass2_incrementer += 1
        if data_dict_columns['Survived'][r] == 1:
            pclass2_survived += 1
    if data_dict_columns['Pclass'][r] == 3:
        pclass3_incrementer += 1
        if data_dict_columns['Survived'][r] == 1:
            pclass3_survived += 1

pclass1_survival_rate = pclass1_survived/pclass1_incrementer
pclass2_survival_rate = pclass2_survived/pclass2_incrementer
pclass3_survival_rate = pclass3_survived/pclass3_incrementer
        
print("The survival rate for the upper class was {}".format(pclass1_survival_rate))
print("The survival rate for the middle class was {}".format(pclass2_survival_rate))
print("The survival rate for the lower class was {}".format(pclass3_survival_rate))

The survival rate for the upper class was 0.6296296296296297
The survival rate for the middle class was 0.47282608695652173
The survival rate for the lower class was 0.24236252545824846


# 11. Independent Work

Use the techniques from step 10 to make different conditional splits in the `Survived` column. Can you find a combination of splits that maximizes the survival rate?

In [68]:
high_class_female_survived = 0
high_class_female_incrementer = 0
for r in range(len(data_dict_columns['Sex'])):
    if data_dict_columns['Sex'][r] == 'female' and data_dict_columns['Pclass'][r] == 1:
        high_class_female_incrementer += 1
        if data_dict_columns['Survived'][r] == 1:
            high_class_female_survived += 1

high_class_female_survival_rate = high_class_female_survived/high_class_female_incrementer

print("The survival rate for high class females was {}".format(high_class_female_survival_rate))

The survival rate for high class females was 0.9680851063829787


# 12. Distributions

### A) For each of your numeric features, use `pyplot` subplots to plot a hisogram for each feature.

* Make sure to title each subplot.
* If you get an error, it may be caused by `np.NaN`

In [69]:
import matplotlib.pyplot as plt
%matplotlib inline

### B) for each of these values, what's the 90% confidence interval of the population mean?

* Create a function to find the confidence interval, and use it on each of the numeric columns.
* What's your interpretation of the interval?

In [70]:
from scipy import stats
# Your code here

# 13. Pandas

### A: Load the titanic csv into a `DataFrame` using `pd.read_csv()`

In [71]:
import pandas as pd

In [72]:
titanic_df = pd.read_csv('titanic.csv')

### B: Display the first 5 rows, the last 4 rows, and a sample of 3 rows.

In [73]:
display(titanic_df.head(5))
display(titanic_df.tail(4))
display(titanic_df.sample(3))

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

PassengerId  Survived  Pclass                                      Name  \
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

PassengerId  Survived  Pclass                 Name   Sex   Age  SibSp  \
363          364         0       3      Asim, Mr. Adola  male  35.0      0   
619          620         0       2  Gavey, Mr. Lawrence  male  26.0      0   
837          838         0       3  Sirota, Mr. Maurice  male   NaN      0   

     Parch              Ticket   Fare Cabin Embarked  
363      0  SOTON/O.Q. 3101310   7.05   NaN        S  
619      0               31028  10.50   NaN        S  
837      0              392092   8.05   NaN        S

### C: Create a row mask that is `True` when `Pclass == 3`. Use this to mask your `DataFrame`. Find the mean of the `Survived` column. Is it the same as what we calculated in part 10?

In [74]:
pclass3_mask = titanic_df['Pclass'] == 3
titanic_df_pclass3 = titanic_df.loc[pclass3_mask]
pclass3_survivors = titanic_df_pclass3['Survived'].sum()
print(pclass3_survivors/491)
print("This is the same survival rate we got in part 10 of this lab. 119 survivors of 491 total passengers belonging to Pclass 3 (low class)")

0.242362525458
This is the same survival rate we got in part 10 of this lab. 119 survivors of 491 total passengers belonging to Pclass 3 (low class)


### D: Using a `.groupby()`, what is the mean of the survival column grouped by `Pclass` and `Sex`. What are your observations?

In [81]:
print(titanic_df.groupby(['Sex', 'Pclass'])['Survived'].mean())

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64


The higher class passengers were more likely to have survived the sinking of the Titanic than the lower class passengers. However, a low class female was still more likely to have survived than a high class male, therefore gender is more of a determining factor than class, however there is a strong correlation between class and survival.

### E: Survival Rate by Age Range:  `pd.cut()` takes two arguments: A `list`, `Series`, or `array`, and a list of bins. Create a new column in your `DataFrame` using `pd.cut()` that groups your ages into bins of 5 years. Then, use `.groupby()` to display the survival rate and count for each age group

In [94]:
new_age = pd.cut(titanic_df['Age'], bins=[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80])
titanic_df.groupby(new_age)['Survived'].mean()

Age
(0, 5]      0.704545
(5, 10]     0.350000
(10, 15]    0.578947
(15, 20]    0.343750
(20, 25]    0.344262
(25, 30]    0.388889
(30, 35]    0.465909
(35, 40]    0.417910
(40, 45]    0.361702
(45, 50]    0.410256
(50, 55]    0.416667
(55, 60]    0.388889
(60, 65]    0.285714
(65, 70]    0.000000
(70, 75]    0.000000
(75, 80]    1.000000
Name: Survived, dtype: float64

# 14. Hypothesis Testing

### A) Hypothesis:

Create a null and alternate hypothesis to ask the following quesiton: Was the `Age` of survivors different from that of people who didn't survive?

**Hypotheses:**

$H_0$: The age of those who did survive was the same, on average, as the age of those who did not survive.

$H_1$: The age of those who did survive was different, on average, from the age of those who did not survive.

### B) T-Testing

Use a t-test to test your null hypothesis. What's the p-value? What's your interpretation? Do you accept or reject your null hypothesis? What does this mean in terms of `Age`?

In [101]:
ttest = stats.ttest_ind

In [133]:
new_titanic_df = titanic_df.loc[titanic_df['Age'].isnull()].index
withoutNull_titanic_df = titanic_df.drop(new_titanic_df)

average_age_notsurvived = withoutNull_titanic_df.loc[titanic_df['Survived'] == 0]['Age']
average_age_survived = withoutNull_titanic_df.loc[titanic_df['Survived'] == 1]['Age']

print("The average age of those who did not survive was {}".format(average_age_notsurvived.mean()))
print("The average age of those who did survive was {}".format(average_age_survived.mean()))

ttest(average_age_survived, average_age_notsurvived)

The average age of those who did not survive was 30.62617924528302
The average age of those who did survive was 28.343689655172415


Ttest_indResult(statistic=-2.0666869462538102, pvalue=0.039124654013482488)

A pvalue of 0.0391 means that there is a 3.91% chance that the average age was the same amongst survivors as it was for non-survivors, and that the difference in mean ages between the two sets was purely by chance. Since the pvalue is less than 0.05, this percentage is too low and therefore I reject the null hypothesis. This means that the age of those who did not survive (30.626 years old) was not the same, on average, as the age of those who did survive (28.344 years old) and the difference is statistically significant.

# 13. Write-up

Use markdown cells to answer the following questions:

1. What is the main difference between a list and a tuple?
2. Can you iterate over a dictionary? If so, how?
3. What is the term for a list or array of `True` and `False` values used to select certain rows or columns in a `DataFrame`?
4. If you have a continuous variable, when would the median be a better descriptor than the mean? Why?
5. Give a qualitative description of the survivors of the titanic based on the effects of your splits in the data on survival rate. Or, given a row, what columns would you look at to guess if they survived?


$1$. A list is mutable and a tuple is immutable.

$2$. Yes, but you must use a for loop. An example is shown in the cell below. Each time through the for loop, the i value is updated to the "next" key in dict1. Using this key, the value is accessed and assigned to our new dictionary (dict2).

In [152]:
dict1 = {1:2, 2:2, 3:5}
dict2 = {}
for i in dict1:
    dict2[i] = dict1[i]

dict2

{1: 2, 2: 2, 3: 5}

$3$. Mask

$4$. If there are 400 values and 399 of them are 4.21 and the other value is 421.21, the median value of 4.21 would describe the data much more accurately than the average of 5.25. In general, when there are large outliers, the median is a better descriptor than the mean.\

$5$. I would first look at the gender of the passenger, and if the passenger was female, that would greatly improve their chances of surviving. The next column I would look at would be the Pclass, and the lower the Pclass is, the greater chance of survival. For example, a female in Pclass 1 has a 96.81% chance of survival, whereas a male in Pclass 3 has a 13.54% chance of survival. Although age is considered statistically significant, it would be hard to deem whether or not a specific age would improve the chances of survival, especially for passengers in the age ranges from 30-60. However, if the passenger was to an extreme side of the age range (0-5 or 80+), this passenger has a very high likelihood of surviving.

# 14. Evaluation

Please use markdown cells to submit your responses. 

1. What was easy for you in this project?
2. What was difficult?
3. Where did you make the most improvement?
4. Where would you like to improve?

$1$. Steps 1-3 and 5-7 were very easy for me since it was very similar to what we did in the Pokemon Lab. I also have some experience using pandas dataframes, so a few of those were simple, however I also learned the groupby() and pd.cut() which are very useful.

$2$. I am struggling to understand how to form dictionary comprehensions correctly. I have always used normal for loops, and it is hard for me to pick up the correct way to use the dictionary comprehensions. I also have trouble dealing with missing/"dirty" data that may be in the wrong form.

$3$. At first, I had never used the groupby() function before and did not really understand how it should work. After using stack overflow and various other resources, I finally started to understand it. I messed around with it for a little, and realized it is a very useful tool for seeing correlations between different data.

The same goes for the pd.cut. I did not understand how to initiate the bin instances, but I figured it out and realized how useful that can be for data analysis.

$4$. I would like to figure out how to use dictionary and list comprehensions. I think this could save a ton of time, because it is much less typing and less lines of code. It cleans up the code and makes it more readable as well.